File: frontend/src/types/workflowVideo.ts
/**
 * Video workflow v1 API-aligned types (W1). Mirrors backend/app/schemas/workflow_video.py.
 */

import type { TemplateCapabilities } from '@/types/api'

export type VideoRunKind = 'batch' | 'production'

export interface VideoApprovedTopic {
  topic_id: string
  title: string
  content?: string | null
  reason?: string | null
  source_submitter_id?: string | null
  source_node_instance_id?: string | null
  script_author_id: string
  due_at?: string | null
}

export interface ParticipantUserPreview {
  id: string
  email: string
  display_name?: string | null
}

export interface TopicCaptureSubmitResponse {
  task_id: string
  node_instance_id: string
  topic_count: number
  topics: Array<{
    topic_id?: string
    title: string
    content?: string | null
    reason?: string | null
  }>
}

export interface InstanceSubmissionsResponse {
  instance_id: string
  node_key: string
  submissions: Array<{
    node_instance_id: string
    node_key: string
    instance_key: string
    assignee_user_id: string | null
    assignee_email?: string | null
    assignee_display_name?: string | null
    submitted_at?: string | null
    topics: TopicCaptureSubmitResponse['topics']
  }>
}

export interface FinalizeTopicsResponse {
  instance_id: string
  approved_count: number
  fork_status: string
  fork_deferred: boolean
  message?: string | null
}

export interface CloseCaptureResponse {
  instance_id: string
  capture_closed: boolean
  capture_closed_at: string
  skipped_capture_count: number
  message: string
}

export interface DispatchTopicRequest {
  topic_id: string
  title: string
  script_writer_user_id: string
  source_node_instance_id?: string | null
}

export interface DispatchTopicResponse {
  instance_id: string
  child_instance_id: string
  fork_status: string
  message?: string | null
}

export interface RejectedCaptureItem {
  reason: string
  topic_id?: string
  instance_key?: string
}

export interface RejectCapturesRequest {
  rejections: RejectedCaptureItem[]
  source_node_key?: string
}

export interface RejectCapturesResponse {
  instance_id: string
  reopened_count: number
  reopened_instance_keys: string[]
}

export interface RejectProductionStepRequest {
  reason: string
  target_node_key?: string | null
}

export interface RejectProductionStepResponse {
  instance_id: string
  target_node_key: string
  target_node_instance_id: string
  iteration: number
}

export interface ForkProductionRunsResponse {
  batch_instance_id: string
  forked_count: number
  skipped_count: number
  child_instance_ids: string[]
  fork_status: string
}

export interface CreateGraphTemplateRunRequest {
  inputs?: Record<string, unknown>
  participants_snapshot: Record<
    string,
    { mode: 'all' | 'subset'; user_ids: string[]; include_initiator?: boolean }
  >
  department_id?: string | null
  run_label?: string | null
}

export interface CreateGraphTemplateRunResponse {
  instance_id: string
  root_task_id: string
  run_kind: string
  activated_task_count: number
  node_instance_count: number
  current_node_key?: string | null
}

export interface PreviewParticipantsResponse {
  policy_ref: string
  mode: 'all' | 'subset'
  user_ids: string[]
  users: ParticipantUserPreview[]
  snapshot: {
    mode: 'all' | 'subset'
    user_ids: string[]
    include_initiator?: boolean
  }
}

export interface LaunchSchemaField {
  key: string
  label: string
  type: 'text' | 'textarea' | 'datetime' | 'user' | string
  required?: boolean
}

export interface LaunchSchema {
  fields: LaunchSchemaField[]
}

export interface CaptureSchemaColumn {
  key: string
  label: string
  type: 'text' | 'textarea' | 'user' | 'datetime' | string
  required?: boolean
  pool_key?: string | null
}

export interface CaptureSchema {
  mode: 'row_table'
  min_rows?: number
  max_rows?: number
  columns: CaptureSchemaColumn[]
}

export interface GraphTemplateSummary {
  id: string
  code: string
  name: string
  description?: string | null
  status: string
  version: number
  /** @deprecated Legacy video v1 product type from config.run_kind. Prefer capabilities. */
  run_kind?: string | null
  tags?: string[]
  capabilities?: TemplateCapabilities
  config?: Record<string, unknown>
  scope_mode?: 'global' | 'departments'
  scope_department_ids?: string[]
  run_count_total?: number | null
  run_count_30d?: number | null
  active_run_count?: number | null
}

export interface GraphTemplateNodeDetail {
  id: string
  node_key: string
  title: string
  sort_order: number
  node_type?: string
  assignment_mode?: string
  join_mode?: string
  routing_mode?: 'exclusive' | 'inclusive' | 'parallel' | 'first_match'
  assignee_rule?: Record<string, unknown>
  config?: Record<string, unknown>
}

export interface GraphTemplateEdgeDetail {
  id?: string | null
  from_node_key: string
  to_node_key: string
  is_reject_path: boolean
  condition?: Record<string, unknown>
  priority?: number
}

export interface GraphTemplateDesignerDetail extends GraphTemplateSummary {
  base_code: string
  source_template_id?: string | null
  has_instances: boolean
  structure_locked: boolean
  nodes: GraphTemplateNodeDetail[]
  edges?: GraphTemplateEdgeDetail[]
}

export interface GraphTemplateValidateResult {
  valid: boolean
  errors: string[]
}

export interface GraphTemplateExportBundle {
  format_version: number
  template: {
    name: string
    description?: string | null
    config?: Record<string, unknown>
    context_schema?: Record<string, unknown>
    scope_mode?: 'global' | 'departments'
    scope_department_ids?: string[]
    nodes: GraphTemplateNodeDetail[]
    edges?: GraphTemplateEdgeDetail[]
  }
}

export interface GraphTemplateDryRunPolicyPreview {
  policy_ref: string
  mode: string
  user_count: number
  user_ids: string[]
}

export interface GraphTemplateDryRunResult {
  valid: boolean
  errors: string[]
  schema_snapshot: Record<string, unknown>
  normalized_inputs: Record<string, unknown>
  entry_node_keys: string[]
  participant_previews: GraphTemplateDryRunPolicyPreview[]
}

export interface WorkflowGraphInstanceSummary {
  id: string
  template_id: string | null
  status: string
  current_node_key?: string | null
  run_label?: string | null
  parent_instance_id?: string | null
  context: Record<string, unknown>
  progress_percent?: number
  total_node_count?: number
  completed_node_count?: number
  active_task_id?: string | null
}

export interface WorkflowRunEventItem {
  id: string
  instance_id: string
  event_type: string
  event_version: number
  aggregate_version: number | null
  command_id: string | null
  causation_id: string | null
  correlation_id: string | null
  actor_user_id: string | null
  payload: Record<string, unknown>
  occurred_at: string
  created_at: string
}

export interface WorkflowRunEventListResponse {
  instance_id: string
  items: WorkflowRunEventItem[]
  total: number
  limit: number
  offset: number
}

export interface DepartmentRunSummary {
  instance_id: string
  run_label: string | null
  status: string
  created_at: string
  event_count: number
  department_id: string | null
}

export interface VideoRunContext {
  run_kind?: VideoRunKind
  run_label?: string | null
  inputs?: Record<string, unknown>
  participants_snapshot?: Record<
    string,
    { mode: 'all' | 'subset'; user_ids: string[] }
  >
  root_task_id?: string | null
  template_version?: number
  schema_snapshot?: Record<string, unknown>
  approved_topics?: VideoApprovedTopic[]
  fork_status?: 'pending' | 'completed' | 'partial'
  parent_instance_id?: string | null
  topic_id?: string | null
  topic_title?: string | null
  script_author_id?: string | null
  forked_child_instance_ids?: string[]
}

export interface GraphTemplateSchedule {
  id: string
  template_id: string
  template_code?: string | null
  template_name?: string | null
  name: string
  scope_department_id: string
  scope_department_name?: string | null
  scope_mode: 'self' | 'subtree'
  cron_expr: string
  timezone: string
  default_inputs: Record<string, unknown>
  run_label_template?: string | null
  participant_mode: 'all' | 'subset'
  participant_user_ids: string[]
  exclude_department_ids: string[]
  exclude_user_ids: string[]
  is_active: boolean
  created_by: string
  next_run_at: string | null
  last_run_at: string | null
  last_run_status: string | null
  last_run_message: string | null
  last_run_instance_count: number | null
  created_at: string
  updated_at: string
}

export interface CreateGraphTemplateScheduleRequest {
  template_id: string
  name: string
  scope_department_id: string
  scope_mode?: 'self' | 'subtree'
  cron_expr: string
  timezone?: string
  default_inputs?: Record<string, unknown>
  run_label_template?: string | null
  participant_mode?: 'all' | 'subset'
  participant_user_ids?: string[]
  exclude_department_ids?: string[]
  exclude_user_ids?: string[]
  is_active?: boolean
}

export interface GraphTemplateScheduleRunNowResponse {
  created_count: number
  skipped_count: number
  failed_count: number
  details: Array<Record<string, unknown>>
}

File: backend/app/schemas/workflow_video.py
"""Video workflow v1 schemas: form engine + run context (W1)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, ValidationError, field_validator, model_validator

RunKind = Literal["batch", "production"]
CaptureColumnType = Literal["text", "textarea", "number", "datetime", "user"]
AggregateOnConfirmAction = Literal["finalize_topics_and_fork", "advance_only"]


class LaunchFieldSchema(BaseModel):
  key: str = Field(min_length=1, max_length=64)
  label: str = Field(min_length=1, max_length=120)
  type: Literal["text", "textarea", "datetime", "user", "user_multi", "department"] = "text"
  required: bool = False
  policy_ref: str | None = Field(default=None, max_length=64)


class LaunchSchema(BaseModel):
  fields: list[LaunchFieldSchema] = Field(default_factory=list, min_length=1)


class CaptureColumnSchema(BaseModel):
  key: str = Field(min_length=1, max_length=64)
  label: str = Field(min_length=1, max_length=120)
  type: CaptureColumnType = "text"
  required: bool = False


class CaptureSchema(BaseModel):
  mode: Literal["row_table"] = "row_table"
  min_rows: int = Field(default=1, ge=1, le=100)
  max_rows: int = Field(default=20, ge=1, le=200)
  columns: list[CaptureColumnSchema] = Field(min_length=1)
  storage: Literal["deliverable_payload"] = "deliverable_payload"
  completion_policy: Literal["on_capture_submitted"] = "on_capture_submitted"

  @model_validator(mode="after")
  def _validate_row_bounds(self) -> CaptureSchema:
    if self.max_rows < self.min_rows:
      raise ValueError("max_rows 不能小于 min_rows。")
    return self


class AggregateAssigneeColumnSchema(BaseModel):
  key: str = Field(min_length=1, max_length=64)
  label: str = Field(min_length=1, max_length=120)
  type: Literal["user"] = "user"


class AcceptanceRejectToSchema(BaseModel):
  node_key: str = Field(min_length=1, max_length=64)
  instance_key_from: Literal["topic_id", "assignee"] | None = None


class AcceptanceSpecSchema(BaseModel):
  reject_to: AcceptanceRejectToSchema | None = None


class AggregateOnConfirmSchema(BaseModel):
  action: AggregateOnConfirmAction
  child_template_code: str | None = Field(default=None, max_length=64)
  idempotency_key: str | None = Field(default=None, max_length=64)


class AggregateSchema(BaseModel):
  mode: Literal["submission_matrix"] = "submission_matrix"
  source_node_key: str = Field(min_length=1, max_length=64)
  row_id_field: str = Field(default="topic_id", max_length=64)
  row_actions: list[Literal["approve", "reject"]] = Field(default_factory=lambda: ["approve", "reject"])
  assignee_column: AggregateAssigneeColumnSchema | None = None
  on_confirm: AggregateOnConfirmSchema

  @model_validator(mode="after")
  def _validate_fork_action(self) -> AggregateSchema:
    if self.on_confirm.action == "finalize_topics_and_fork":
      if not self.on_confirm.child_template_code:
        raise ValueError("finalize_topics_and_fork 必须配置 child_template_code。")
      if not self.on_confirm.idempotency_key:
        raise ValueError("finalize_topics_and_fork 必须配置 idempotency_key。")
    return self


class ParticipantPolicyDefinition(BaseModel):
  type: Literal["department_members"] = "department_members"
  department_id: UUID | None = None
  exclude_initiator_by_default: bool = True


class ParticipantsSnapshotEntry(BaseModel):
  mode: Literal["all", "subset"]
  user_ids: list[UUID] = Field(default_factory=list)
  include_initiator: bool = False

  @model_validator(mode="after")
  def _validate_subset(self) -> ParticipantsSnapshotEntry:
    if self.mode == "subset" and not self.user_ids:
      raise ValueError("subset 模式至少选择一名参与人。")
    return self


class ApprovedTopic(BaseModel):
  topic_id: UUID
  title: str = Field(min_length=1, max_length=255)
  content: str | None = None
  reason: str | None = None
  source_submitter_id: UUID | None = None
  source_node_instance_id: UUID | None = None
  script_author_id: UUID
  due_at: str | None = None


class TopicCaptureRow(BaseModel):
  model_config = ConfigDict(extra="allow")

  topic_id: UUID | None = None
  title: str | None = Field(default=None, max_length=255)
  content: str | None = None
  reason: str | None = None


class TopicCaptureSubmitRequest(BaseModel):
  topics: list[TopicCaptureRow] = Field(default_factory=list)


class TopicCaptureSubmitResponse(BaseModel):
  task_id: UUID
  node_instance_id: UUID
  topic_count: int
  topics: list[TopicCaptureRow]


class NodeSubmissionRead(BaseModel):
  node_instance_id: UUID
  node_key: str
  instance_key: str
  assignee_user_id: UUID | None
  assignee_email: str | None = None
  assignee_display_name: str | None = None
  submitted_at: datetime | None = None
  topics: list[TopicCaptureRow] = Field(default_factory=list)


class InstanceSubmissionsResponse(BaseModel):
  instance_id: UUID
  node_key: str
  submissions: list[NodeSubmissionRead]


class RejectedCaptureItem(BaseModel):
  reason: str = Field(min_length=1, max_length=2000)
  topic_id: UUID | None = None
  instance_key: str | None = Field(default=None, max_length=64)

  @model_validator(mode="after")
  def _validate_target(self) -> RejectedCaptureItem:
    has_topic = self.topic_id is not None
    has_instance = bool(self.instance_key and str(self.instance_key).strip())
    if has_topic == has_instance:
      raise ValueError("必须且只能指定 topic_id 或 instance_key 之一。")
    return self


class RejectCapturesRequest(BaseModel):
  rejections: list[RejectedCaptureItem] = Field(min_length=1)
  source_node_key: str = Field(default="N1_PROPOSE", min_length=1, max_length=64)


class RejectCapturesResponse(BaseModel):
  instance_id: UUID
  reopened_count: int
  reopened_instance_keys: list[str]


class RejectProductionStepRequest(BaseModel):
  reason: str = Field(min_length=1, max_length=2000)
  target_node_key: str | None = Field(default=None, max_length=64)


class RejectProductionStepResponse(BaseModel):
  instance_id: UUID
  target_node_key: str
  target_node_instance_id: UUID
  iteration: int


class FinalizeTopicsRequest(BaseModel):
  approved_topics: list[ApprovedTopic] = Field(default_factory=list)
  rejected_topics: list[RejectedCaptureItem] = Field(default_factory=list)

  @model_validator(mode="after")
  def _validate_finalize_payload(self) -> FinalizeTopicsRequest:
    if not self.approved_topics and not self.rejected_topics:
      raise ValueError("approved_topics 与 rejected_topics 不能同时为空。")
    return self


class ForkProductionRunsResponse(BaseModel):
  batch_instance_id: UUID
  forked_count: int
  skipped_count: int
  child_instance_ids: list[UUID]
  fork_status: str


class FinalizeTopicsResponse(BaseModel):
  instance_id: UUID
  approved_count: int
  fork_status: str
  fork_deferred: bool = False
  child_instance_ids: list[UUID] = Field(default_factory=list)
  message: str | None = None


class DispatchTopicRequest(BaseModel):
  topic_id: UUID
  title: str = Field(min_length=1, max_length=255)
  script_writer_user_id: UUID
  source_node_instance_id: UUID | None = None


class DispatchTopicResponse(BaseModel):
  instance_id: UUID
  child_instance_id: UUID
  fork_status: str
  message: str | None = None


class CreateGraphTemplateRunRequest(BaseModel):
  inputs: dict[str, Any] = Field(default_factory=dict)
  participants_snapshot: dict[str, ParticipantsSnapshotEntry] = Field(min_length=1)
  department_id: UUID | None = None
  run_label: str | None = Field(default=None, max_length=255)


class CreateGraphTemplateRunResponse(BaseModel):
  instance_id: UUID
  root_task_id: UUID
  run_kind: str
  activated_task_count: int
  node_instance_count: int
  current_node_key: str | None = None


class PreviewParticipantsRequest(BaseModel):
  mode: Literal["all", "subset"] = "all"
  user_ids: list[UUID] = Field(default_factory=list)
  department_id: UUID | None = None


class ParticipantUserPreview(BaseModel):
  id: UUID
  email: str
  display_name: str | None = None


class PreviewParticipantsResponse(BaseModel):
  policy_ref: str
  mode: Literal["all", "subset"]
  user_ids: list[UUID]
  users: list[ParticipantUserPreview]
  snapshot: ParticipantsSnapshotEntry


class WorkflowRunEventRead(BaseModel):
  model_config = ConfigDict(from_attributes=True)

  id: UUID
  instance_id: UUID
  event_type: str
  actor_user_id: UUID | None
  payload: dict[str, Any] = Field(default_factory=dict)
  created_at: datetime


class WorkflowRunEventListResponse(BaseModel):
  instance_id: UUID
  items: list[WorkflowRunEventRead]
  total: int
  limit: int
  offset: int


class WorkflowRunContextSchema(BaseModel):
  """Validated subset of graph instance context for video v1."""

  model_config = ConfigDict(extra="allow")

  run_kind: RunKind | None = None
  run_label: str | None = None
  inputs: dict[str, Any] = Field(default_factory=dict)
  participants_snapshot: dict[str, ParticipantsSnapshotEntry] = Field(default_factory=dict)
  root_task_id: UUID | None = None
  template_version: int | None = Field(default=None, ge=1)
  schema_snapshot: dict[str, Any] = Field(default_factory=dict)
  approved_topics: list[ApprovedTopic] = Field(default_factory=list)
  fork_status: Literal["pending", "completed", "partial"] | None = None
  parent_instance_id: UUID | None = None
  topic_id: UUID | None = None
  topic_title: str | None = None
  script_author_id: UUID | None = None
  edit_assignee_id: UUID | None = None
  publish_at: str | None = None
  platform: str | None = None
  publish_title: str | None = None
  archived: bool | None = None
  archived_at: str | None = None
  forked_child_instance_ids: list[UUID] = Field(default_factory=list)


class WorkflowGraphTemplateNodeConfigSchema(BaseModel):
  """Node config including graph execution + form engine (W1)."""

  model_config = ConfigDict(extra="allow")

  kind: Literal["single", "multi_instance"] = "single"
  expand_from: str | None = Field(default=None, max_length=64)
  participant_policy_ref: str | None = Field(default=None, max_length=64)
  launch_schema: LaunchSchema | None = None
  capture_schema: CaptureSchema | None = None
  aggregate_schema: AggregateSchema | None = None
  acceptance_spec: AcceptanceSpecSchema | None = None
  handshake_required: bool = False
  completion_policy: str | None = None

  @model_validator(mode="after")
  def _validate_multi_instance(self) -> WorkflowGraphTemplateNodeConfigSchema:
    if self.kind == "multi_instance" and not self.expand_from:
      raise ValueError("multi_instance 节点必须配置 expand_from。")
    if self.capture_schema is not None and self.aggregate_schema is not None:
      raise ValueError("同一节点不能同时配置 capture_schema 与 aggregate_schema。")
    return self


def validate_launch_schema(payload: dict[str, Any]) -> LaunchSchema:
  return LaunchSchema.model_validate(payload)


def validate_capture_schema(payload: dict[str, Any]) -> CaptureSchema:
  return CaptureSchema.model_validate(payload)


def validate_aggregate_schema(payload: dict[str, Any]) -> AggregateSchema:
  return AggregateSchema.model_validate(payload)


def validate_node_config(payload: dict[str, Any]) -> WorkflowGraphTemplateNodeConfigSchema:
  return WorkflowGraphTemplateNodeConfigSchema.model_validate(payload)


def validate_run_context(payload: dict[str, Any]) -> WorkflowRunContextSchema:
  return WorkflowRunContextSchema.model_validate(payload)


__all__ = [
  "AggregateOnConfirmAction",
  "AggregateSchema",
  "ApprovedTopic",
  "CaptureSchema",
  "CreateGraphTemplateRunRequest",
  "CreateGraphTemplateRunResponse",
  "LaunchSchema",
  "ParticipantPolicyDefinition",
  "ParticipantUserPreview",
  "ParticipantsSnapshotEntry",
  "PreviewParticipantsRequest",
  "PreviewParticipantsResponse",
  "RunKind",
  "WorkflowGraphTemplateNodeConfigSchema",
  "WorkflowRunContextSchema",
  "WorkflowRunEventListResponse",
  "WorkflowRunEventRead",
  "validate_aggregate_schema",
  "validate_capture_schema",
  "validate_launch_schema",
  "validate_node_config",
  "validate_run_context",
  "ValidationError",
]
